packages/components/form/src/types.ts

```diff
@@ -19,7 +20,55 @@ export type FormLabelWidthContext = ReturnType<typeof useFormLabelWidth>
 export interface FormItemRule extends RuleItem {
   trigger?: Arrayable<string>
 }
-export type FormRules = Partial<Record<string, Arrayable<FormItemRule>>>
+
+// Reference https://stackoverflow.com/questions/58434389/typescript-deep-keyof-of-a-nested-object/58436959
+// Concatenates two strings with a dot in the middle, unless the last string is empty
+// Join<"a","b"> => "a.b"
+// Join<"a",""> => "a"
+type Join<K, P> = K extends string | number
+  ? P extends string | number
+    ? `${K}${'' extends P ? '' : '.'}${P}`
+    : never
+  : never
+// Limit the recursion to avoid typescript error
+type Prev = [
+  never,
+  0,
+  1,
+  2,
+  3,
+  4,
+  5,
+  6,
+  7,
+  8,
+  9,
+  10,
+  11,
+  12,
+  13,
+  14,
+  15,
+  16,
+  17,
+  18,
+  19,
+  20,
+  ...0[]
+]
+type Leaves<T, D extends number = 10> = [D] extends [never]
+  ? never
+  : T extends object
+  ? { [K in keyof T]-?: Join<K, Leaves<T[K], Prev[D]>> }[keyof T]
+  : ''
+export type FormRules<
+  T extends MaybeRef<Record<string, any> | string> = string
+> = Partial<
+  Record<
+    UnwrapRef<T> extends string ? UnwrapRef<T> : Leaves<UnwrapRef<T>>,
+    Arrayable<FormItemRule>
+  >
+>
```

Comment:
IMO, Perhaps we should only support simple types, using MaybeRef seems redundant. what do you think?

example

```ts
interface Form {
  a: number
}

const form = ref<Form>() // or reactive<Form>()

const formRules: FormRules<Form> = {
  a: { ... },
}
```